# Data Science & Solutions Technical Assessment

##### Going to import Pandas because it is a well-suited library for analyzing datasets like shots data

In [1]:
import pandas as pd

##### Read in the shots data csv file

In [2]:
df = pd.read_csv("shots_data.csv")

##### Print the dataframe

In [3]:
df

team     x     y  fgmade
0    Team A -23.1   3.5       0
1    Team A   0.0  25.1       1
2    Team A   0.5   1.0       1
3    Team A  -5.6   5.2       0
4    Team A   4.0  14.5       1
..      ...   ...   ...     ...
555  Team B   7.8  24.7       0
556  Team B  10.1   1.4       1
557  Team B -17.8  -0.1       0
558  Team B  22.6   2.9       0
559  Team B  -0.7  14.7       1

[560 rows x 4 columns]

#### This method can determine the shot zone based on the coordinates of where the shot was taken

In [4]:
def shot_zone(x, y):
    if (abs(x) >= 22.0) and (y <= 7.8):
        return "C3"
    elif ((abs(x) <= 23.75) and (y > 7.8)) and y > 23.75:
        return "NC3"
    else:
        return "2PT"

##### Apply the method

In [5]:
df["shot_zone"] = df.apply(lambda row: shot_zone(row.x, row.y), axis=1)

##### New dataframe including shot zones

In [6]:
df

team     x     y  fgmade shot_zone
0    Team A -23.1   3.5       0        C3
1    Team A   0.0  25.1       1       NC3
2    Team A   0.5   1.0       1       2PT
3    Team A  -5.6   5.2       0       2PT
4    Team A   4.0  14.5       1       2PT
..      ...   ...   ...     ...       ...
555  Team B   7.8  24.7       0       NC3
556  Team B  10.1   1.4       1       2PT
557  Team B -17.8  -0.1       0       2PT
558  Team B  22.6   2.9       0        C3
559  Team B  -0.7  14.7       1       2PT

[560 rows x 5 columns]

# Team A

In [7]:
team_a_stats = df.loc[df["team"] == "Team A"]

In [8]:
team_a_stats

team     x     y  fgmade shot_zone
0    Team A -23.1   3.5       0        C3
1    Team A   0.0  25.1       1       NC3
2    Team A   0.5   1.0       1       2PT
3    Team A  -5.6   5.2       0       2PT
4    Team A   4.0  14.5       1       2PT
..      ...   ...   ...     ...       ...
275  Team A  23.0   4.7       1        C3
276  Team A  -3.0  25.4       0       NC3
277  Team A  -8.0   2.6       1       2PT
278  Team A  -7.6   1.7       0       2PT
279  Team A  -5.5   6.2       1       2PT

[280 rows x 5 columns]

In [9]:
team_a_distrib = team_a_stats.groupby("shot_zone").count() /len(team_a_stats)

In [10]:
team_a_distrib

team         x         y    fgmade
shot_zone                                        
2PT        0.778571  0.778571  0.778571  0.778571
C3         0.071429  0.071429  0.071429  0.071429
NC3        0.150000  0.150000  0.150000  0.150000

In [11]:
team_a_distrib = team_a_distrib.drop(["team", "x", "y"],axis = 1)

### **This is the percentage of team A shots within each zone**

In [12]:
team_a_distrib

fgmade
shot_zone          
2PT        0.778571
C3         0.071429
NC3        0.150000

### Now we are going to find the `effective field goal %` for each shot zone for **Team A**

### **Corner Threes**

In [13]:
corner_threes = team_a_stats[team_a_stats["shot_zone"] == "C3"]

In [14]:
len(corner_threes)

20

In [15]:
team_a_corner_makes = int(len(corner_threes[corner_threes["fgmade"] == 1]))

In [16]:
team_a_corner_makes

8

In [17]:
c3_eff_fg_percentage = (team_a_corner_makes + (0.5 * team_a_corner_makes))/ int(len(corner_threes))

In [18]:
c3_eff_fg_percentage

0.6

#### Once we do this calculation,  we can conclude that the effective field goal percentage for Team A for Corner 3s is `60%`

### **Non-Corner Threes**

In [19]:
team_a_nc3s = team_a_stats[team_a_stats["shot_zone"] == "NC3"]

In [20]:
len(team_a_nc3s)

42

In [21]:
team_a_nc3_makes = int(len(team_a_nc3s[team_a_nc3s["fgmade"] == 1]))

In [22]:
team_a_nc3_makes

15

In [23]:
nc3_eff_fg_percentage = (team_a_nc3_makes + (0.5 * team_a_nc3_makes))/ int(len(team_a_nc3s))

In [24]:
nc3_eff_fg_percentage

0.5357142857142857

#### Once we do this calculation,  we can conclude that the effective field goal percentage for Team A for Non-Corner 3s is `53.6%`

### **2PT**

In [25]:
team_a_2pt_attempts = team_a_stats[team_a_stats["shot_zone"] == "2PT"]

In [26]:
len(team_a_2pt_attempts)

218

In [27]:
team_a_2pt_makes = int(len(team_a_2pt_attempts[team_a_2pt_attempts["fgmade"] == 1]))

In [28]:
team_a_2pt_makes

82

In [29]:
two_pt_eff_fg_percentage = team_a_2pt_makes / int(len(team_a_2pt_attempts))

In [30]:
two_pt_eff_fg_percentage

0.3761467889908257

#### Once we do this calculation, we can conclude that the effective field goal percentage for Team A for 2PTs is `37.6%`

## Team B

In [31]:
team_b_stats = df.loc[df["team"] == "Team B"]

In [32]:
team_b_stats

team     x     y  fgmade shot_zone
280  Team B   2.5   6.2       0       2PT
281  Team B   7.4  18.3       1       2PT
282  Team B   1.1   0.1       0       2PT
283  Team B  -1.6  25.6       0       NC3
284  Team B -14.4  21.7       1       2PT
..      ...   ...   ...     ...       ...
555  Team B   7.8  24.7       0       NC3
556  Team B  10.1   1.4       1       2PT
557  Team B -17.8  -0.1       0       2PT
558  Team B  22.6   2.9       0        C3
559  Team B  -0.7  14.7       1       2PT

[280 rows x 5 columns]

In [33]:
team_b_distrib =  team_b_stats.groupby("shot_zone").count() / len(team_b_stats)

In [34]:
team_b_distrib = team_b_distrib.drop(["team", "x", "y"],axis = 1)

### This is the percentage of team B shots within each zone

In [35]:
team_b_distrib

fgmade
shot_zone          
2PT        0.789286
C3         0.075000
NC3        0.135714

### Now we are going to find the `effective field goal %` for each shot zone for **Team B**

## **Corner Threes**

In [36]:
team_b_corner_threes = team_b_stats[team_b_stats["shot_zone"] == "C3"]

In [37]:
len(team_b_corner_threes)

21

In [38]:
team_b_corner_makes = int(len(team_b_corner_threes[team_b_corner_threes["fgmade"] == 1]))

In [39]:
team_b_corner_makes

5

In [40]:
c3_eff_fg_percentage = (team_b_corner_makes + (0.5 * team_b_corner_makes))/ int(len(team_b_corner_threes))

In [41]:
c3_eff_fg_percentage

0.35714285714285715

#### The effective field goal percentage for **corner threes** is `35.7%` for Team B

## Non-Corner Threes

In [42]:
team_b_nc3s = team_b_stats[team_b_stats["shot_zone"] == "NC3"]

In [43]:
len(team_b_nc3s)

38

In [44]:
team_b_nc3_makes = int(len(team_b_nc3s[team_b_nc3s["fgmade"] == 1]))

In [45]:
team_b_nc3_makes

13

In [46]:
nc3_eff_fg_percentage = (team_b_nc3_makes + (0.5 * team_b_nc3_makes)) / int(len(team_b_nc3s))

In [47]:
nc3_eff_fg_percentage

0.5131578947368421

#### Once we do this calculation, we can conclude that the effective field goal percentage for Team B for Non_Corner 3s is 51.3%

## 2PT

In [48]:
team_b_2pt_attempts = team_b_stats[team_b_stats["shot_zone"] == "2PT"]

In [49]:
team_b_2pt_attempts

team     x     y  fgmade shot_zone
280  Team B   2.5   6.2       0       2PT
281  Team B   7.4  18.3       1       2PT
282  Team B   1.1   0.1       0       2PT
284  Team B -14.4  21.7       1       2PT
285  Team B -10.6  23.5       1       2PT
..      ...   ...   ...     ...       ...
553  Team B   1.1   2.3       1       2PT
554  Team B  20.5  14.1       0       2PT
556  Team B  10.1   1.4       1       2PT
557  Team B -17.8  -0.1       0       2PT
559  Team B  -0.7  14.7       1       2PT

[221 rows x 5 columns]

In [50]:
team_b_2pt_makes = int(len(team_b_2pt_attempts[team_b_2pt_attempts["fgmade"] == 1]))

In [51]:
team_b_2pt_makes

97

In [52]:
two_pt_eff_fg_percentage = team_b_2pt_makes / int(len(team_b_2pt_attempts))

In [53]:
two_pt_eff_fg_percentage

0.43891402714932126

### Once we do this calculation, we can conclude that the effective field goal percentage for Team B's 2PT conversions is 43.9%